src/libraries/System.Private.CoreLib/src/System/ComAwareWeakReference.cs

```diff
@@ -0,0 +1,179 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Threading;
+
+#if FEATURE_COMINTEROP || FEATURE_COMWRAPPERS
+
+using static System.WeakReferenceHandleTags;
+
+namespace System
+{
+    internal sealed partial class ComAwareWeakReference
+    {
+        private readonly nint _weakHandle;
+        private ComInfo? _comInfo;
+
+        internal sealed class ComInfo
+        {
+            private readonly IntPtr _pComWeakRef;
+            private readonly long _wrapperId;
+
+            internal object? ResolveTarget()
+            {
+                return ComWeakRefToObject(_pComWeakRef, _wrapperId);
+            }
+
+            // see: syncblk.h
+            private const int IS_HASHCODE_BIT_NUMBER = 26;
+            private const int BIT_SBLK_IS_HASHCODE = 1 << IS_HASHCODE_BIT_NUMBER;
+            private const int BIT_SBLK_IS_HASH_OR_SYNCBLKINDEX = 0x08000000;
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            internal static unsafe ComInfo? FromObject(object? target)
+            {
+                if (target == null)
+                    return null;
+
+                fixed (byte* pRawData = &target.GetRawData())
+                {
+                    // The header is 4 bytes before MT field on all architectures
+                    int header = *(int*)(pRawData - sizeof(IntPtr) - sizeof(int));
+                    // common case: target does not have a syncblock, so there is no interop info
+                    if ((header & (BIT_SBLK_IS_HASH_OR_SYNCBLKINDEX | BIT_SBLK_IS_HASHCODE)) != BIT_SBLK_IS_HASH_OR_SYNCBLKINDEX)
+                        return null;
+                }
+
+                return FromObjectSlow(target);
+            }
+
+            [MethodImpl(MethodImplOptions.NoInlining)]
+            private static unsafe ComInfo? FromObjectSlow(object target)
+            {
+                IntPtr pComWeakRef = ObjectToComWeakRef(target, out long wrapperId);
+                if (pComWeakRef == 0)
+                    return null;
+
+                return new ComInfo(pComWeakRef, wrapperId);
```

Comment:
This will leak `pComWeakRef` reference when `new ComInfo` fails (very rare).

Reply:
yes, could be worth wrapping in try/catch - just in case there is OOM or something like that.

Reply:
Do we need to care about a possibility of asynchronous exception (like a ThreadAbort) _after_ the instance is created?

Reply:
I think we generally do not harden code against ThreadAbort any more.

Reply:
We do not care about thread abort.